Integrate go perun tutorial

matthiasgeihs opened the pull request.

ggwpez commented on source/conf.py:
```diff
@@ -19,9 +19,9 @@
-copyright = u'2020, Hyperledger'
-author = u'Hyperledger'
+project = u'Perun Framework'
+copyright = u'2020, Perun Framework Contributors'
```
2021.  
Regarding the contributors: Maybe we also need a `NOTICE` file here? @sebastianst

choeppler replied:
Actually, the docs project started last year, so 2020 is okay IMO.

matthiasgeihs replied:
Isn't the copyright year usually updated?

choeppler replied:
It doesn't really matter that much. Some add years whenever there were substantive changes, i.e you'd procede from `2020` via `2020, 2021` to `2020--2022`...

matthiasgeihs replied:
Thank you for your comments.

@choeppler I adjusted the copyright holder in the footer accordingly.

choeppler commented:
The CircleCI build broke because of some invalid external links. You can also run that test with `make linkcheck` locally.

ggwpez commented:
The CI check for the tutorial code is now missing. @matthiasgeihs

ggwpez commented:
> The CircleCI build broke because of some invalid external links. You can also run that test with `make linkcheck` locally.

Weird since the link that is reported as broken works fine. Maybe it is inserted after the page-load into the DOM.

choeppler commented on source/go-perun/index.rst:
```diff
@@ -0,0 +1,49 @@
+Go-Perun
+========
+
+`go-perun <https://github.com/perun-network/go-perun>`_ is a Go implementation of the Perun state channel protocols.
```
```suggestion
`go-perun <https://github.com/hyperledger-labs/go-perun>`_ is a Go implementation of the Perun state channel protocols.
```

ggwpez replied:
We did not review the links in all our projects yet to point to `HLL` instead of `perun-network`. This discussion is a good reminder.

ggwpez commented on source/go-perun/index.rst:
```diff
@@ -0,0 +1,49 @@
+Go-Perun
```
There are now multiple top-level sites:  
`Introduction`, `Concepts`, `Go-Perun` and `Perun-node`.  
I think this is a bit confusion to someone who opens the page for the first time. We should at least rename `Go-Perun` to `Tutorial`.

matthiasgeihs replied:
I think this has to do how sphinx documentation pages work. The idea is that go-perun can cover anything related to the go-perun Framework component, not just the tutorial.

manoranjith commented:
From the CI logs, these were the broken links:
1. `(go-perun/tutorial/channels/updating: line   33) broken    disputes.html -`

As @ggwpez  pointed out, this link is be valid but is reported as broken, probably since it is inserted after page load.
(go-perun/tutorial/getting-started: line   35) broken    https://github.com/trufflesuite/ganache-cli#installation - Anchor 'installation' not found

manoranjith commented on source/go-perun/tutorial/channels/updating.rst:
```diff
@@ -0,0 +1,41 @@
+
+.. _UpdateBy: https://pkg.go.dev/perun.network/go-perun/client#Channel.UpdateBy
+.. _Update: https://pkg.go.dev/perun.network/go-perun/client#Channel.Update
+.. _dispute process: disputes.html
```
```suggestion
```
Change 1/4 for fixing broken link issue.

Use sphinx anchor links instead of external links. This ensures the links work even when the documentation is built to formats other than html (such as PDF).

manoranjith commented on source/go-perun/tutorial/channels/updating.rst:
```diff
@@ -0,0 +1,41 @@
+   :language: go
+   :lines: 117-124
+
+An update can also be rejected with a reason. This starts the `dispute process`_.
```
```suggestion
An update can also be rejected with a reason. This starts the :ref:`dispute process<Disputes>`.
```
Change 2/4 for fixing broken link issue. Update link text.

manoranjith commented on source/go-perun/tutorial/channels/disputes.rst:
```diff
@@ -0,0 +1,21 @@
+Disputes
```
```suggestion
.. _Disputes:

Disputes
```

Change 3/4 for fixing broken link issue. Add custom label.

manoranjith commented on source/go-perun/tutorial/setup/contracts.rst:
```diff
@@ -0,0 +1,81 @@
+we always restart the *ganache-cli* before running it.
+
+.. _ValidateAssetHolderETH: https://pkg.go.dev/perun.network/go-perun/backend/ethereum/channel#ValidateAssetHolderETH
+.. _dispute: ../channels/disputes.html
```
```suggestion
```
Change 4/4 to fix broken link.

This link anchor is not used anywhere in this document. In case, if you need to link to the dispute page, you can use this format directly, along with a label in the dispute page.

```
ref:`dispute process<Disputes>`
```
where `dispute process` can be replaces with the text that you want to display.

manoranjith reviewed (commented):
Suggestions to fix broken link. Includes 4 changes.

matthiasgeihs commented:
> The CircleCI build broke because of some invalid external links. You can also run that test with `make linkcheck` locally.



> From the CI logs, these were the broken links:
> 
> 1. `(go-perun/tutorial/channels/updating: line   33) broken    disputes.html -`
> 
> I have added a suggestion with 4 changes to fix this one.
> 
> As @ggwpez pointed out, this link is be valid but is reported as broken, probably since it is inserted after page load.
> 
> 1. (go-perun/tutorial/getting-started: line   35) broken    https://github.com/trufflesuite/ganache-cli#installation - Anchor 'installation' not found

any idea how to fix this? @manoranjith

manoranjith commented:
> > The CircleCI build broke because of some invalid external links. You can also run that test with `make linkcheck` locally.
> 
> 
> 
> > From the CI logs, these were the broken links:
> > 
> > 1. `(go-perun/tutorial/channels/updating: line   33) broken    disputes.html -`
> > 
> > I have added a suggestion with 4 changes to fix this one.
> > 
> > As @ggwpez pointed out, this link is be valid but is reported as broken, probably since it is inserted after page load.
> > 
> > 1. (go-perun/tutorial/getting-started: line   35) broken    https://github.com/trufflesuite/ganache-cli#installation - Anchor 'installation' not found
> 
> any idea how to fix this? @manoranjith 

For the first one, have made a suggestion, for the second one I haven't found a fix yet, will have to still look for it.

In case, if we are not able to find any fix for it,y suggestion would be make `linkcheck` as a separate CI job and make it optional. The reviewer will then have to decide whether the Pr can be merged in cases where the job fails.

choeppler commented:
> In case, if we are not able to find any fix for it,y suggestion would be make `linkcheck` as a separate CI job and make it optional. The reviewer will then have to decide whether the Pr can be merged in cases where the job fails.

I dimly remember that there's a timeout option for the linkecheck. I think I once saw some other docs projects where that was increased. Could that be a solution for the problem?

manoranjith commented:
> > In case, if we are not able to find any fix for it,y suggestion would be make `linkcheck` as a separate CI job and make it optional. The reviewer will then have to decide whether the Pr can be merged in cases where the job fails.
> 
> I dimly remember that there's a timeout option for the linkecheck. I think I once saw some other docs projects where that was increased. Could that be a solution for the problem?

I tried the timeout option with `30s`. It still was reported as broken link.

Another option is to exclude check of anchors in this. In this case, the validity of link will be checked. I will add a suggestion to make the required changes in the `conf.py` file. @matthiasgeihs  @choeppler

choeppler commented:
@manoranjith, now that you've investigated this linkcheck problem and we have an example, could you also file a bug report with the sphinx project to maybe get the root cause solved? Thanks!

manoranjith commented:
> 
> 
> @manoranjith, now that you've investigated this linkcheck problem and we have an example, could you also file a bug report with the sphinx project to maybe get the root cause solved? Thanks!

I think this is a known limitation in `linkcheck`, because they have provided options exclusively to handle these scenarios where anchors in the URL are dynamically added (say by Javascript). However, I can check the existing issues in their repo to see the understand more and will create an issue if there is none covering this problem.

manoranjith commented on source/index.rst:
```diff
@@ -1,16 +1,35 @@
+
+|
+
+The Perun Framework
```
@matthiasgeihs  What are your thoughts on merging the `Introduction` section in `introduction` section here. Because both of the sections seem to target the same purpose of providing first glance introduction to the project.

The `introduction.rst` page will then begin with the `Perun Protocol` as its first section.

matthiasgeihs replied:
I thought about it too. But it is some work to do it right. If we move the introduction part from there, you may want to write a different header for that part and potentially reorganize the section a bit accordingly. For example, I could see that we merge "Introduction" and "Concepts" into one larger section.

Anyhow, I think applying such structural changes is some work and I would propose to do it in another PR.

manoranjith commented on source/index.rst:
```diff
@@ -1,16 +1,35 @@
 Table of contents
 -----------------
 .. toctree::
+   :maxdepth: 2
```
@matthiasgeihs  This is a good one. :) Because after adding `go-perun` the toc tree in the landing page will be too long with all levels of sub-headings.

manoranjith commented on source/go-perun/index.rst:
```diff
@@ -0,0 +1,48 @@
+Go-Perun
+========
+
```
Could we add a short introduction to `go-perun` before the `developer tutorial` section ?

matthiasgeihs replied:
Yes, this was also planned. But it needs some time and I propose to do it in another PR.

matthiasgeihs commented:
I have addressed all comments, I think. Please have another look.

I externalized the tutorial source code into a separate repository at perun-network, where we can also gather additional code examples. The code is included here via a git submodule. I added submodule initialization to the ci and added a remark to the README.

Please have a look.

ggwpez commented:
We could also add `https://github.com/perun-network/perun-tutorial` as a sub-module here instead of copying the files.  
But i was not involved in the planning meeting for the integration, so consider this void if it does not make sense.

manoranjith commented:
@matthiasgeihs The PR LGTM now. I have 2 minor comments and a review from @choeppler on the change in `conf.py`: 
- one is a suggestion to add (in the readme file) the steps for checking out git sub modules
- other, as a remainder to create follow up issues.

- Need review comments from @choeppler  Removing authors and changing the copyright holder name from `Hyperledger` to `Perun Framework Contributors` in `conf.py`.

The PR can be merged after this.

manoranjith commented:
@matthiasgeihs  Also could you add a one-line explanation in this commit message a09bfabab262d87c9722cbc624a4ab97e5865faf, the reason for disabling `autosection extension`. The reason, I understand is to use manually created tags instead.

choeppler commented on source/go-perun/tutorial/index.rst:
```diff
@@ -0,0 +1,25 @@
+
+This tutorial describes how to build a simple payment channel application on top of Ethereum using `go-perun <https://github.com/hyperledger-labs/go-perun>`_.
+The tutorial targets developers familiar with Golang and Ethereum.
+The source code is available at `GitHub <https://github.com/perun-network/perun-tutorial>`_.
```
If you're including the source code from a third-party under a different license, then the license obligations need to be fulfilled. Provide license, NOTICE file if available, ...
```suggestion
The source code snippets within this tutorial are literally included from *perun-tutorial*, (c) PolyCrypt GmbH 2021, licensed under `Apache-2.0 <https://github.com/perun-network/perun-tutorial/blob/master/LICENSE>`_. The entire source code is available at `GitHub <https://github.com/perun-network/perun-tutorial>`_.
```

matthiasgeihs replied:
ok

choeppler commented on source/conf.py:
```diff
@@ -19,9 +19,9 @@
-copyright = u'2020, Hyperledger'
-author = u'Hyperledger'
+project = u'Perun Framework'
+copyright = u'2020, Perun Framework Contributors'
+author = u'Perun Framework Contributors'
```
@tkuhrt, @ryjones: Is there a recommendation from you on how copyright statements of Hyperledger (Labs) documentation should look like? The way this is all set up, we have common ownership as no copyright is transferred. So the proposed change fits quite well IMO, even though, strictly speaking, there can be and quite often is a difference between copyrights-holders and authors... 

@matthiasgeihs , IIRC, [_templates/footer.html](https://github.com/hyperledger-labs/perun-doc/blob/master/source/_templates/footer.html) needs also to be patched accordingly.

ryjones replied:
We (Hyperledger) has been fairly liberal on this point. We're more concerned that the license (Apache 2, CC-4-BY) is correctr.

tkuhrt replied:
As Ry mentioned, we have been fairly liberal; however, there is this [wiki page](https://wiki.hyperledger.org/display/TSC/Copyright+and+License+Policy) that talks about Copyright and License Policy.

matthiasgeihs commented:
> We could also add `https://github.com/perun-network/perun-tutorial` as a sub-module here instead of copying the files.
> But i was not involved in the planning meeting for the integration, so consider this void if it does not make sense.

I thought about it, but there were several changes necessary, hence I did not follow this approach.

matthiasgeihs commented on README.md:
```diff
@@ -14,6 +14,8 @@ HTML documentation is published at
 
 ## Build
 
+Please ensure that the repository is checked out with submodules.
+
```
Hm, in some cases, I prefer cooperative over collaborative. I think it's fine to have not have it consistent across the whole doc as long as it is consistent within each part.

choeppler commented on source/_templates/footer.html:
```diff
@@ -18,7 +18,7 @@
         {% trans path=pathto('copyright'), copyright=copyright|e %}<a href="{{ path }}">Copyright</a> &copy; {{ copyright }}.{% endtrans %}
         {%- else %}
-        {% trans copyright=copyright|e %}Copyright &copy; 2020 Hyperledger {% endtrans %}
+        {% trans copyright=copyright|e %}Copyright &copy; 2021 Perun Framework Contributors {% endtrans %}
```
```suggestion
        {% trans copyright=copyright|e %}Copyright &copy; 2020 Perun Framework Contributors {% endtrans %}
```
or
```suggestion
        {% trans copyright=copyright|e %}Copyright &copy; 2020, 2021 Perun Framework Contributors {% endtrans %}
```

matthiasgeihs replied:
OK, changed the copyright year to 2020-2021.

manoranjith reviewed (approved):
LGTM.
